codalab/server/bundle_manager.py

```diff
@@ -445,11 +439,19 @@ def get_sort_key(worker):
             else:
                 deps = set(worker['dependencies'])
                 num_available_deps = len(needed_deps & deps)
-            worker_id = worker['worker_id']
 
-            # if the bundle doesn't request GPUs (only request CPUs), prioritize workers that don't have GPUs
-            gpu_priority = bundle_resources.gpus or not has_gpu[worker_id]
-            return (gpu_priority, num_available_deps, worker['cpus'], random.random())
+            # GPU priority is useful when sorting bundles that request for only CPUs. In this case, we want
+            # to rank CPU workers as high as possible and GPU workers as low as possible, so that GPU workers
+            # could be better utilized towards GPU oriented jobs.
+            gpu_priority = worker['gpus'] == 0
```

Comment:
Generalize this to `-worker['gpus']` (want to favor workers with fewer GPUs).
Also prioritize workers with fewer CPUs by the same general logic (choose the worker with the fewest resources that still satisfies the resource requirements)?

Reply:
Yes. I have added `worker['cpus']` right after the shared dependencies, which will prioritize the number of shared dependencies first and then `worker['cpus']`. We could also move its position higher, depending on how important we think those dependencies are. Let me know if we want to change the order.